source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,228 @@
+uint64_t reflectionHashField(const Protobuf::Message& message,
+                             const Protobuf::FieldDescriptor& field, uint64_t seed);
+
+// To make a map serialize deterministically we need to ignore the order of
+// the map fields. To do that, we simply combine the hashes of each entry
+// using an unordered operator (addition), and then apply that combined hash to
+// the seed.
+uint64_t reflectionHashMapField(const Protobuf::Message& message,
+                                const Protobuf::FieldDescriptor& field, uint64_t seed) {
+  const Protobuf::Reflection& reflection = *message.GetReflection();
+  ASSERT(field.is_map());
+  const auto& entries = reflection.GetRepeatedFieldRef<Protobuf::Message>(message, &field);
+  ASSERT(!entries.empty());
+  const Protobuf::Descriptor& map_descriptor = *entries.begin()->GetDescriptor();
+  const Protobuf::FieldDescriptor& key_field = *map_descriptor.map_key();
+  const Protobuf::FieldDescriptor& value_field = *map_descriptor.map_value();
+  uint64_t combined_hash = 0;
+  for (const Protobuf::Message& entry : entries) {
+    uint64_t entry_hash = reflectionHashField(entry, key_field, 0);
+    entry_hash = reflectionHashField(entry, value_field, entry_hash);
+    combined_hash += entry_hash;
+  }
+  return HashUtil::xxHash64Value(combined_hash, seed);
+}
+
+uint64_t reflectionHashField(const Protobuf::Message& message,
+                             const Protobuf::FieldDescriptor& field, uint64_t seed) {
+  using Protobuf::FieldDescriptor;
+  const Protobuf::Reflection& reflection = *message.GetReflection();
+  seed = HashUtil::xxHash64Value(field.number(), seed);
+  switch (field.cpp_type()) {
+  case FieldDescriptor::CPPTYPE_INT32:
+    seed = hashScalarField<int32_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_UINT32:
+    seed = hashScalarField<uint32_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_INT64:
+    seed = hashScalarField<int64_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_UINT64:
+    seed = hashScalarField<uint64_t>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_DOUBLE:
+    seed = hashScalarField<double>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_FLOAT:
+    seed = hashScalarField<float>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_BOOL:
+    seed = hashScalarField<bool>(reflection, message, field, seed);
+    break;
+  case FieldDescriptor::CPPTYPE_ENUM:
+    if (field.is_repeated()) {
+      int c = reflection.FieldSize(message, &field);
+      for (int i = 0; i < c; i++) {
+        seed = HashUtil::xxHash64Value(reflection.GetRepeatedEnumValue(message, &field, i), seed);
+      }
+    } else {
+      seed = HashUtil::xxHash64Value(reflection.GetEnumValue(message, &field), seed);
+    }
+    break;
+  case FieldDescriptor::CPPTYPE_STRING: {
```

Comment:
nit: seems that the other "case" places don't have the outer curly brackets. Let's make this consistent (at least in this function).

Reply:
Ah, yes, that was a "leak" from the scratch string being declared, but it ended up inside an if-branch so the outer braces were no longer needed.

Reply:
fwiw I use curlies in a switch only when I need a local var; I haven't attempted to then change all the cases to match that in the past within Envoy.